api/src/test/java/org/openmrs/ObsTest.java

```diff
@@ -554,4 +646,272 @@ public void setFormField_shouldRejectAPathContainingTheSeparator() throws Except
 		Obs obs = new Obs();
 		obs.setFormField("", path);
 	}
+	
+	/**
+	 * @see Obs#isDirty()
+	 * @verifies return false when no change has been made
+	 */
+	@Test
+	public void isDirty_shouldReturnFalseWhenNoChangeHasBeenMade() throws Exception {
+		assertFalse(new Obs().isDirty());
+		
+		//Should also work if setters are called with same values as the original
+		Obs obs = createObs();
+		obs.setGroupMembers(new LinkedHashSet<>());
+		obs.getConcept().setDatatype(new ConceptDatatype());
+		assertFalse(obs.isDirty());
+		BeanUtils.copyProperties(obs, BeanUtils.cloneBean(obs));
+		assertFalse(obs.isDirty());
+	}
+	
+	/**
+	 * @see Obs#isDirty()
+	 * @verifies return true when any immutable field has been changed
+	 */
+	@Test
+	public void isDirty_shouldReturnTrueWhenAnyImmutableFieldHasBeenChanged() throws Exception {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Method names should not contain underscores](https://www.codacy.com/app/openmrs/openmrs-core/file/2318734355/issues/source?bid=3204099&fileBranchId=3204127#l672)